### PR TITLE
optimise upload blob path and report format

### PR DESCRIPTION
**Pull request by ruowan**
add code severity

support load resource type

output all file to outputFolder

### CODE REVIEW

File: lib/testScenario/dataMasker.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    }
+    return ret;
+  }
+}
+
+/* ref: https://stackoverflow.com/questions/1144783/how-to-replace-all-occurrences-of-a-string-in-javascript
+ * doc: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#escaping
+ */
+function escapeRegExp(string: string) {
+  return string.replace(/[.*+\-?^${}()|[\]\\]/g, "\\$&"); // $& means the whole matched string
```
Comment (raych1) [inline lib/testScenario/dataMasker.ts:132]: Should regex ignore case too?

Comment (ruowan) [reply lib/testScenario/dataMasker.ts:132]: no

File: lib/testScenario/postmanReportParser.ts [hunk 2]
```diff
@@ -34,30 +35,26 @@ export interface NewmanReportParserOption extends FileLoaderOption {
...
-    const content = await this.fileLoader.load(this.opts.newmanReportFilePath);
-    const report = JSON.parse(content) as NewmanReport;
-    this.report.metadata.testScenarioFilePath = report.collection.info.description.content; // JSON.parse(report.collection.description.content);
-    for (const it of report.run.executions) {
-      this.report.executions.push(this.generateExampleItem(it));
+  public async generateRawReport(newmanReportFilePath: string) {
+    const ret: RawReport = { variables: {}, executions: [], timings: {}, metadata: {} };
+    const content = await this.fileLoader.load(newmanReportFilePath);
+    const newmanReport = JSON.parse(content) as NewmanReport;
+    ret.metadata = JSON.parse(newmanReport.collection.info.description.content); // JSON.parse(report.collection.description.content);
```
Comment (raych1) [inline lib/testScenario/postmanReportParser.ts:48]: Is the schema changed? Can you remove the commented code?

File: lib/testScenario/reportGenerator.ts [hunk 3]
```diff
@@ -153,8 +217,36 @@ export class ReportGenerator {
...
+          this.dataMasker.jsonStringify(this.swaggerExampleQualityResult)
+        );
+
+        await this.blobUploader.uploadContent(
+          "reportforpipeline",
+          blobPath,
+          this.dataMasker.jsonStringify(this.swaggerExampleQualityResult)
+        );
+
+        for (const [correlationId, v] of this.recording) {
```
Comment (raych1) [inline lib/testScenario/reportGenerator.ts:242]: Is payload masked the sensitive data?

File: lib/testScenario/dataMasker.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+  public jsonParse(content: string): any {
+    try {
+      return JSON.parse(this.maskString(content));
+    } catch (err) {
+      return undefined;
+    }
+  }
+
+  public maskObject(obj: any, addMaskedValue = false): any {
```
Comment (raych1) [inline lib/testScenario/dataMasker.ts:66]: Can you add comment about mask logic?

File: lib/testScenario/dataMasker.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    return ret;
+  }
+
+  public maybeSecretKey(key: any): boolean {
+    if (typeof key !== "string") {
+      return false;
+    }
+    return (
+      this.maskKeys.some((it) => key.toLowerCase().includes(it)) ||
+      this.maskValuePatterns.some((it) => it.test(key))
```
Comment (raych1) [inline lib/testScenario/dataMasker.ts:102]: Why here test the value while this function seems to check key?

File: lib/testScenario/dataMasker.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  maskValue: MaskFunction;
+}
+
+export const defaultMaskValue = (_content?: string): string => {
+  return "<masked>";
+};
+
+@injectable()
+export class DataMasker {
+  public maskKeys: string[] = [
```
Comment (raych1) [inline lib/testScenario/dataMasker.ts:22]: Should maskKeys be lower case strings?

Comment (ruowan) [reply lib/testScenario/dataMasker.ts:22]: No. But macth function is case insensitive.

File: lib/testScenario/reportGenerator.ts (hunk 3 above)
Comment (ruowan) [reply lib/testScenario/reportGenerator.ts:242]: Good catch. Mask payload.

File: lib/testScenario/dataMasker.ts (hunk 4 above)
Comment (ruowan) [reply lib/testScenario/dataMasker.ts:66]: Sure. Added

File: lib/testScenario/reportGenerator.ts [hunk 7]
```diff
@@ -15,51 +23,79 @@ import {
...
-  };
-  stepResult: { [step: string]: StepResult };
+  testScenarioFilePath: string;
+  readmeFilePath?: string;
+  swaggerFilePaths: string[];
+  tag?: string;
+
+  // New added fields
+  environment?: string;
+  armEnv?: string;
```
[the remark this answers is not in the document]
Comment (ruowan) [reply lib/testScenario/reportGenerator.ts:47]: I am not sure about how to identify armEnv.  Can armEnv be identify by location or other endpoint?

File: lib/testScenario/reportGenerator.ts [hunk 8]
```diff
@@ -252,6 +401,19 @@ export class ReportGenerator {
...
 
+  private getRuntimeError(it: RawExecution): RuntimeError {
+    const ret: RuntimeError = {
+      code: "",
+      message: "",
+      severity: "Error",
+      detail: this.dataMasker.jsonStringify(it.response.body),
+    };
+    const responseObj = this.dataMasker.jsonParse(it.response.body);
+    ret.code = responseObj?.error?.code;
```
Comment (raych1) [inline lib/testScenario/reportGenerator.ts:412]: Is it possible to set with these values?
code = `RUNTIME_ERROR_${statusCode}`,
message = "code:xxx, message:xxx".

File: lib/testScenario/reportGenerator.ts [hunk 9]
```diff
@@ -69,43 +105,64 @@ export class ReportGenerator {
...
+      providerNamespace: getProviderFromFilePath(this.opts.testDefFilePath),
+      apiVersion: getApiVersionFromSwaggerFile(this.opts.swaggerFilePaths![0]),
+      runId: this.opts.runId,
+      rootPath: this.fileRoot,
+      repository: process.env.SPEC_REPOSITORY,
+      branch: process.env.SPEC_BRANCH,
+      commitHash: process.env.COMMIT_HASH,
+      environment: process.env.ENVIRONMENT || "test",
+      testScenarioName: this.opts.testScenarioName,
+      armEndpoint: "https://management.azure.com",
```
[the remark this answers is not in the document]
Comment (ruowan) [reply lib/testScenario/reportGenerator.ts:148]: Yes. Will added in next PR.

File: lib/testScenario/reportGenerator.ts [hunk 10]
```diff
@@ -409,7 +409,7 @@ export class ReportGenerator {
       detail: this.dataMasker.jsonStringify(it.response.body),
     };
     const responseObj = this.dataMasker.jsonParse(it.response.body);
-    ret.code = "RUNTIME_ERROR";
+    ret.code = `RUNTIME_ERROR_${it.response.statusCode}`;
     ret.message = responseObj?.error?.message;
```
Comment (raych1) [inline lib/testScenario/reportGenerator.ts:413]: Can you set message = code:${error?.code}, message:${error?.message}

Comment (ruowan) [reply lib/testScenario/reportGenerator.ts]: Updated.

**Review by raych1 — APPROVED** [review]

**Pull request merged**